packages/core/src/routes/applications/application-secret.ts

```diff
@@ -0,0 +1,106 @@
+import { Applications, ApplicationSecrets, internalPrefix } from '@logto/schemas';
+import { generateStandardSecret } from '@logto/shared';
+import { z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import koaGuard from '#src/middleware/koa-guard.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import { type ManagementApiRouter, type RouterInitArgs } from '../types.js';
+
+export const generateInternalSecret = () => internalPrefix + generateStandardSecret();
+
+export default function applicationSecretRoutes<T extends ManagementApiRouter>(
+  ...[router, { queries }]: RouterInitArgs<T>
+) {
+  router.delete(
+    '/applications/:id/legacy-secret',
+    koaGuard({
+      params: z.object({ id: z.string().min(1) }),
+      response: Applications.guard,
+      status: [200, 400, 404],
+    }),
+    async (ctx, next) => {
+      const {
+        params: { id },
+      } = ctx.guard;
+
+      const application = await queries.applications.findApplicationById(id);
+
+      if (!application.secret || application.secret.startsWith(internalPrefix)) {
+        throw new RequestError('application.no_legacy_secret_found');
+      }
+
+      const secret = generateInternalSecret();
+      ctx.body = await queries.applications.updateApplicationById(id, { secret });
+      return next();
+    }
+  );
```

Comment:
Implementation in this API seems DB migration related things, why don't we write a DB alteration script instead?

Reply:
we should keep backward compatibility, otherwise all current secrets will be invalid